### PR TITLE
[FIX] Custom Oauth login not working with accessToken

### LINKED ISSUE #14108: Custom Oauth login with provider access Token
We have a method called: `registerAccessTokenService` 
Source found: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/oauth.js#L9

It is executed from the login endpoint: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/oauth.js#L27

So when you do a login like: 
```
curl -H "Content-type:application/json" \
      http://localhost:3000/api/v1/login \
      -d '{ "serviceName": "yourOauth", "accessToken": "tokenFromOauthProvider", "expiresIn": 200 }'
```

serviceName matches the oauth service registered and executes the function passed.  That returns an object like:
```
{
  serviceData,
  options: {
    profile: {
      name: identity.name,
    },
  },
};
```

That then finds or creates an account here: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/oauth.js#L54

An example of an oauth provider registering:
https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/google.js#L28

It takes the access token and calls google with it to get the identity: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/google.js#L10

This is ensuring its valid as well as getting some info about the user that the token belongs to.

The data is then pieced together: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/google.js#L39

picked of only the fields we want: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/lib/server/oauth/google.js#L46

then returned as mentioned above.

That needs added in here: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/custom-oauth/server/custom_oauth_server.js making use of the access patterns setup.  Can likely reuse a lot of the code/functions already there to make this not so bad.

This would also close out: https://github.com/RocketChat/Rocket.Chat/issues/5521

### PR DESCRIPTION
**Pull request by knrt10**
TODO
- [x] Check for whitelist fields, somehow make it configurable maybe asking in curl request as an object.

Closes #14108 

cc @geekgonecrazy would you please review

### CODE REVIEW

File: app/custom-oauth/server/custom_oauth_server.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
+/* eslint-disable complexity */
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]: lets take this off please :)

Comment (knrt10) [reply app/custom-oauth/server/custom_oauth_server.js]: ok

**Comment by knrt10** [comment]
@geekgonecrazy I had to add another field for `access_token` cause different services use different `params` to get identity. Like in many they use `access_token` but in case of gitlab they use `private_token`. Added functionality for this. Please review.

**Comment by ralfbecker** [comment]
I was about to open a similar pull request: https://github.com/RocketChat/Rocket.Chat/commit/63f6e52fc8b01449dd2cd392ed7e0a129283ee47
Also using getIdentity to verify the accessToken, though I added an other function/request to get the lifetime/expiration of the token, as that was always missing from options. So I used RFC7662 OAuth 2.0 Token Introspection, to query it from the server:
```
function getTokeninfo(idToken, config) {
	try {
		const introspectPath = '/introspect';	// not yet configurable in Rocket.Chat, thought RFC defines that path
		return HTTP.post(
			config.serverURL + introspectPath,
			{
				auth: `${ config.clientId }:${ OAuth.openSecret(config.secret) }`,
				headers: {
					Accept: 'application/json',
				},
				params: {
					token: idToken,
					token_type_hint: 'access_token',
				},
			}).data;
	} catch (err) {
		throw _.extend(new Error(`Failed to fetch tokeninfo from custom OAuth ${ config.service }. ${ err.message }`), { response: err.response });
	}
}
```
That of cause requires the server to implement that OAuth 2 extension.

I will try this pull request and report back, if that allows me to access Rocket.Chat Api with accessToken from my server.

Ralf

**Comment by geekgonecrazy** [comment]
Not on introspection.. it for sure will need to be configurable: https://www.oauth.com/oauth2-servers/token-introspection-endpoint/  here they are using /token_info 

in the rfc they use: /introspect - https://tools.ietf.org/html/rfc7662 

Maybe have a setting.. if "" then don't use introspect since the oauth provider probably hasn't implemented it

**Comment by geekgonecrazy** [comment]
@knrt10 made a few adjustments it now works :)

File: app/lib/server/methods/addOAuthService.js [hunk 2]
```diff
@@ -28,7 +28,8 @@ Meteor.methods({
...
 		settings.add(`Accounts_OAuth_Custom-${ name }-identity_path`          , '/me'             , { type: 'string' , group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Identity_Path', persistent: true });
 		settings.add(`Accounts_OAuth_Custom-${ name }-authorize_path`         , '/oauth/authorize', { type: 'string' , group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Authorize_Path', persistent: true });
-		settings.add(`Accounts_OAuth_Custom-${ name }-scope`                  , 'openid'          , { type: 'string' , group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Scope', persistent: true });
+		settings.add(`Accounts_OAuth_Custom-${ name }-scope`, 'openid', { type: 'string', group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Scope', persistent: true });
+		settings.add(`Accounts_OAuth_Custom-${ name }-param_name`, 'access_token', { type: 'string', group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Param_Name', persistent: true });
```
Comment (geekgonecrazy) [inline app/lib/server/methods/addOAuthService.js]: Should we maybe rename this to tokenName or something a bit more obvious?

Comment (knrt10) [reply app/lib/server/methods/addOAuthService.js]: Cool, let me update it

File: app/lib/server/methods/addOAuthService.js [hunk 3]
```diff
@@ -29,7 +29,7 @@ Meteor.methods({
 		settings.add(`Accounts_OAuth_Custom-${ name }-identity_path`          , '/me'             , { type: 'string' , group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Identity_Path', persistent: true });
 		settings.add(`Accounts_OAuth_Custom-${ name }-authorize_path`         , '/oauth/authorize', { type: 'string' , group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Authorize_Path', persistent: true });
 		settings.add(`Accounts_OAuth_Custom-${ name }-scope`, 'openid', { type: 'string', group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Scope', persistent: true });
-		settings.add(`Accounts_OAuth_Custom-${ name }-param_name`, 'access_token', { type: 'string', group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Param_Name', persistent: true });
+		settings.add(`Accounts_OAuth_Custom-${ name }-access_token_param`, 'access_token', { type: 'string', group: 'OAuth', section: `Custom OAuth: ${ name }`, i18nLabel: 'Accounts_OAuth_Custom_Access_Token_Param', persistent: true });
```
Comment (geekgonecrazy) [inline app/lib/server/methods/addOAuthService.js]: can you align these with the comma's above?

Comment (knrt10) [reply app/lib/server/methods/addOAuthService.js]: ok

**Comment by geekgonecrazy** [comment]
This is looking good 👍 i'll give it another pass.

@ralfbecker if you get a chance definitely give this a go and see if works for your case as well

**Comment by ralfbecker** [comment]
A quick update: this pull request does not fix my issue which Rocket.Chat Api access with an accessToken and custom OAuth. It seems the custom OAuth does NOT get registered in AccessTokenServices, it it's called via the Api.
I will try to create an app/lib/server/oauth/custom.js, as in my planned pull request https://github.com/EGroupware/Rocket.Chat/commit/63f6e52fc8b01449dd2cd392ed7e0a129283ee47, but only register knrt10 custom OAuth there.
Any idea how it should supply that, I can't commit to this pull request. I could make an other pull request to knrt10 issue14108 branch.
I'll keep you posted.

**Comment by knrt10** [comment]
@ralfbecker can you specify your process to access the API using token?

**Comment by knrt10** [comment]
Did you follow this step? I am giving an example for github. You need to go to, admin, Oauth, create a new oauth and just fill the URL endpoint and identity input box, like for github URL is `https://api.github.com` and identity is `/user`

. Now go and create access token from your github account and use it in your request like this
```bash
curl -H "Content-type:application/json" \
  http://localhost:3000/api/v1/login \
  -d '{ "serviceName": "yourServiceName", "accessToken": "token","expiresIn": 200 }
```
Also your `yourServiceName` will be one with your new oauth name, also enable it before using it

**Comment by ralfbecker** [comment]
I believe I do exactly the same. Config in Rocket.Chat is like in my [README](https://github.com/EGroupware/openid/blob/master/README.md#rocketchat-custom-oauth-configuration). And I get the exception from here: https://github.com/knrt10/Rocket.Chat/blob/issue14108/app/lib/server/oauth/oauth.js#L31
If I inspect AccessTokenServices in the debugger, only facebook, google and twitter from that directory are registered.
Ralf

**Comment by knrt10** [comment]
It should work, did you rebased the latest changes? I mean some changes were made quite recently

**Comment by knrt10** [comment]
@ralfbecker Its working, I found the error, even if you set name as `EGroupware` it will be converted to lowercase, so you should try like this

```bash
curl -H "Content-type:application/json" \
  http://localhost:3000/api/v1/login \
  -d '{ "serviceName": "egroupware", "accessToken": "token","expiresIn": 200 }'
```
Notice name is `egroupware` not `EGroupware`. Proof of working

Working

Access token is wrong so error

**Comment by ralfbecker** [comment]
That is not the problem, I already use "egroupware" as serviceName.

I just verified, that - for some reason - I made a mistake in merging your pull-request.

I redo it and let you know.

Ralf

**Comment by knrt10** [comment]
Sure, no problem we need testing for this feature, thank you for helping ✌️

**Comment by ralfbecker** [comment]
Ok, I can make now requests to Rocket.Chat Api with my own servers access-tokens :)

I need to check if my pull request https://github.com/RocketChat/Rocket.Chat/pull/14121 is still necessary and if the refreshToken is used.

**Comment by geekgonecrazy** [comment]
Awesome!  Glad you were able to test and get it working!

File: app/custom-oauth/server/custom_oauth_server.js [hunk 4]
```diff
@@ -134,7 +141,7 @@ export class CustomOAuth {
 		}
 	}
 
-	getIdentity(accessToken) {
+	getIdentity(accessToken, accessTokenParam) {
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
	getIdentity(accessToken) {
```

File: app/custom-oauth/server/custom_oauth_server.js [hunk 5]
```diff
@@ -143,7 +150,7 @@ export class CustomOAuth {
 		if (this.identityTokenSentVia === 'header') {
 			headers.Authorization = `Bearer ${ accessToken }`;
 		} else {
-			params.access_token = accessToken;
+			params[accessTokenParam] = accessToken;
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
			params[this.accessTokenParam] = accessToken;
```

File: app/custom-oauth/server/custom_oauth_server.js [hunk 6]
```diff
@@ -173,105 +180,110 @@ export class CustomOAuth {
 		const self = this;
 		OAuth.registerService(this.name, 2, null, (query) => {
 			const accessToken = self.getAccessToken(query);
-			// console.log 'at:', accessToken
 
-			let identity = self.getIdentity(accessToken);
+			const identity = self.getIdentity(accessToken, this.accessTokenParam);
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/custom_oauth_server.js]:
```suggestion
			const identity = self.getIdentity(accessToken);
```

**Comment by charafsalmi** [comment]
Hi,

Is it possible to do it now ? I mean, use the rest api with a keycloak jwt token ?

Thanks,

**Pull request merged**